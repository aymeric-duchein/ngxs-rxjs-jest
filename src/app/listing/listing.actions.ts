import {Issue, SortKey} from '../core/issue/issue.model';

export class ListingIssueClick {
  static readonly type = '[LISTING] Issue click';

  constructor(public issue: Issue) {
  }
}

export class SearchFormUpdated {
  static readonly type = '[LISTING] Search updated';

  constructor(public search: string) {
  }
}

export class SortOptionClick {
  static readonly type = '[LISTING] Sort option click';

  constructor(public sortKey: SortKey) {
  }
}
