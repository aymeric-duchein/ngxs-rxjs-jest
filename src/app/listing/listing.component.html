<div class="search-container">
  <app-search></app-search>
</div>
<div class="sort-container">
  <app-sort></app-sort>
</div>
<app-table></app-table>

<app-add></app-add>
