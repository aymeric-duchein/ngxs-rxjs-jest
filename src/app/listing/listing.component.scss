app-add {
  position: absolute;
  bottom: 1rem;
  right: 3rem;
}


.search-container {
  padding: 2rem 8rem 0.5rem 7rem;
}

.sort-container {
  display: flex;
  padding: 0 8rem;
  justify-content: flex-end;

  app-sort {
    margin-right: 5px;
  }
}

