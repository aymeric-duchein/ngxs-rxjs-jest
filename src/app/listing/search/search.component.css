mat-form-field {
  margin-left: 1em;
  width: calc(100% - 2em);
}
