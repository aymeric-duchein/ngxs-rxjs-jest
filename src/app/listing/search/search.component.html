<form>
  <mat-form-field>
    <input [formControl]="searchFormControl" matInput placeholder="Search">
    <mat-icon matSuffix>search</mat-icon>
  </mat-form-field>
</form>
