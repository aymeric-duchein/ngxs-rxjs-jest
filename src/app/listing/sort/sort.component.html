<button mat-button [matMenuTriggerFor]="menu">Sort by
  <mat-icon>sort</mat-icon>
</button>
<mat-menu #menu="matMenu">
  <button mat-menu-item (click)="sortOptionClick(sortKey.DATE)"> Date
    <ng-container *ngIf="sortCriteria$ | async as sortCriteria">
      <mat-icon *ngIf="sortCriteria.by === sortKey.DATE">
        {{ sortCriteria.direction ? 'arrow_drop_up' : 'arrow_drop_down'}}
      </mat-icon>
    </ng-container>
  </button>
  <button mat-menu-item (click)="sortOptionClick(sortKey.NAME)"> Name
    <ng-container *ngIf="sortCriteria$ | async as sortCriteria">
      <mat-icon *ngIf="sortCriteria.by === sortKey.NAME">
        {{ sortCriteria.direction ? 'arrow_drop_up' : 'arrow_drop_down'}}
      </mat-icon>
    </ng-container>

  </button>
  <button mat-menu-item (click)="sortOptionClick(sortKey.STATUS)">Status
    <ng-container *ngIf="(sortCriteria$ | async) as sortCriteria">

      <mat-icon *ngIf="sortCriteria.by === sortKey.STATUS">
        {{ sortCriteria.direction ? 'arrow_drop_up' : 'arrow_drop_down'}}
      </mat-icon>
    </ng-container>

  </button>
  <button mat-menu-item (click)="sortOptionClick(sortKey.NONE)"> None</button>
</mat-menu>

