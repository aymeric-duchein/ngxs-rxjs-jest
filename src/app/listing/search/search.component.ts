import {Component, OnDestroy, OnInit} from '@angular/core';
import {FormControl} from '@angular/forms';
import {Store} from '@ngxs/store';
import {SearchFormUpdated} from '../listing.actions';
import {Subject} from 'rxjs';
import {debounceTime, takeUntil} from 'rxjs/operators';

@Component({
  selector: 'app-search',
  templateUrl: './search.component.html',
  styleUrls: ['./search.component.css']
})
export class SearchComponent implements OnInit, OnDestroy {

  searchFormControl: FormControl = new FormControl('');
  isDestroyed$: Subject<boolean> = new Subject<boolean>();

  constructor(private readonly store: Store) {
  }

  ngOnInit() {
    this.searchFormControl.valueChanges.pipe(
      debounceTime(300),
      takeUntil(this.isDestroyed$)
    ).subscribe((value) => {
      this.store.dispatch(new SearchFormUpdated(value));
    });
  }

  ngOnDestroy(): void {
    this.isDestroyed$.next(true);
    this.isDestroyed$.complete();
  }
}
