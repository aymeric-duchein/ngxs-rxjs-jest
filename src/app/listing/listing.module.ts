import {NgModule} from '@angular/core';
import {CommonModule} from '@angular/common';
import {ListingComponent} from './listing.component';
import {SearchComponent} from './search/search.component';
import {SortComponent} from './sort/sort.component';
import {TableComponent} from './table/table.component';
import {MatButtonModule, MatFormFieldModule, MatIconModule, MatInputModule, MatMenuModule} from '@angular/material';
import {FormsModule, ReactiveFormsModule} from '@angular/forms';
import {IssueModule} from '../shared/issue/issue.module';
import {AddModule} from '../shared/add/add.module';

@NgModule({
  declarations: [ListingComponent, SearchComponent, SortComponent, TableComponent],
  imports: [
    CommonModule,
    MatFormFieldModule,
    MatIconModule,
    MatInputModule,
    ReactiveFormsModule,
    FormsModule,
    MatButtonModule,
    MatMenuModule,
    IssueModule,
    AddModule,
  ]
})
export class ListingModule {
}
