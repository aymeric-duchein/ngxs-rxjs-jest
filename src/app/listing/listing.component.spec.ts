import {ComponentFixture, TestBed} from '@angular/core/testing';

import {ListingComponent} from './listing.component';
import {MockComponent, MockModule} from 'ng-mocks';
import {SearchComponent} from '@src/app/listing/search/search.component';
import {SortComponent} from '@src/app/listing/sort/sort.component';
import {TableComponent} from '@src/app/listing/table/table.component';
import {AddModule} from '@src/app/shared/add/add.module';

describe('ListingComponent', () => {
  let component: ListingComponent;
  let fixture: ComponentFixture<ListingComponent>;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [ListingComponent, MockComponent(SearchComponent), MockComponent(SortComponent), MockComponent(TableComponent)],
      imports: [MockModule(AddModule)]
    });

    fixture = TestBed.createComponent(ListingComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  test('should create', () => {
    expect(component).toBeTruthy();
    expect(fixture).toMatchSnapshot();
  });
});
