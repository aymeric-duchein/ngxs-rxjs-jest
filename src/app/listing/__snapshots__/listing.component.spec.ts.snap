// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListingComponent should create 1`] = `
<app-listing>
  <div
    class="search-container"
  >
    <app-search />
  </div><div
    class="sort-container"
  >
    <app-sort />
  </div><app-table /><app-add />
</app-listing>
`;
